src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1078 @@
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_oid);
+        val = bsonVal.get_oid();
+    }
+
+    /**
+     * Loads an ObjectID from the current node.
+     *
+     * @param val
+     *    The oid variable into which the OID will be loaded.
+     */
+    void loadValue(bsoncxx::oid& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_oid);
+        val = bsonVal.get_oid().value;
+    }
+
+    /**
+     * Loads a bool from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the bool will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_bool& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_bool);
+        val = bsonVal.get_bool();
+    }
+
+    /**
+     * Loads a BSON datetime from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the datetime will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_date& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_date);
+        val = bsonVal.get_date();
+    }
+
+    /**
+     * Loads a 32 bit int from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the int will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_int32& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_int32);
+        val = bsonVal.get_int32();
+    }
+
+    /**
+     * Loads a 64 bit int from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the int will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_int64& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_int64);
+        val = bsonVal.get_int64();
+    }
+
+    /**
+     * Loads a BSON datetime from the current node.
+     *
+     * @param val
+     *    The time_point variable into which the datetime will be loaded.
+     */
+    void loadValue(std::chrono::system_clock::time_point& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_date);
+
+        val = std::chrono::system_clock::time_point(
+            std::chrono::milliseconds{bsonVal.get_date().value});
+    }
+
+    /**
+     * Loads a bool from the current node.
+     *
+     * @param val
+     *    The boolean variable into which the bool will be loaded.
+     */
+    void loadValue(bool& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_bool);
+        val = bsonVal.get_bool().value;
+    }
+
+    /**
+     * Loads a 32 bit int from the current node.
+     *
+     * @param val
+     *    The int32_t variable into which the int will be loaded.
+     */
+    void loadValue(std::int32_t& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_int32);
+        val = bsonVal.get_int32().value;
```

Comment:
It's unfortunate that these types can't all be covered with a template. The line above could be done by instead calling `BsonT::type_id`, but this line I'm not sure. @hanumantmk - any thoughts?